src/users/user.rs

```diff
@@ -0,0 +1,148 @@
+use std::collections::HashMap;
+use snafu::ensure;
+use pwhash::bcrypt;
+use uuid::Uuid;
+use serde::{Serialize, Deserialize};
+
+use crate::error::Result;
+use crate::error;
+use core::fmt;
+use std::str::FromStr;
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct UserRegistration {
+    pub email: String,
+    pub password: String,
+    pub real_name :String,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct UserCredentials {
+    pub email: String,
+    pub password: String,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct UserIdentification {
+    id: Uuid,
+}
+
+impl UserIdentification {
+    fn new() -> Self {
+        Self { id: Uuid::new_v4() }
+    }
+}
+
+#[derive(Clone)]
+struct User {
+    id: UserIdentification,
+    email: String,
+    password_hash: String,
+    real_name: String,
+    active: bool,
+}
+
+impl User {
+    fn from_user_registration(user_registration: &UserRegistration) -> Self {
```

Comment:
You should implement the From or Into traits instead that do this conversion.
Despite that, I suggest giving the `user_registration` by value because you clone everything and the caller can then decide if it should be copied or moved.

Reply:
8a9b7b5